core/controllers/reader.py

```diff
@@ -120,32 +132,89 @@ def classify(exp_id, state, answer, params):
         state.interaction.id)
     normalized_answer = interaction_instance.normalize_answer(answer)
 
-    # Find the first group that satisfactorily matches the given answer. This
-    # is done by ORing (maximizing) all truth values of all rules over all
-    # answer groups. The group with the highest truth value is considered the
-    # best match.
     best_matched_answer_group = None
     best_matched_answer_group_index = len(state.interaction.answer_groups)
+    best_matched_rule_spec = None
     best_matched_rule_spec_index = None
-    best_matched_truth_value = 0.0
+    best_matched_truth_value = rule_domain.CERTAIN_FALSE_VALUE
+    matched_rule_type = None
+
+    fs = fs_domain.AbstractFileSystem(fs_domain.ExplorationFileSystem(exp_id))
+    input_type = interaction_instance.answer_type
+
+    # Find the first hard rule that matches
     for (answer_group_index, answer_group) in enumerate(
             state.interaction.answer_groups):
-        fs = fs_domain.AbstractFileSystem(
-            fs_domain.ExplorationFileSystem(exp_id))
-        input_type = interaction_instance.answer_type
-        ored_truth_value = 0.0
-        best_rule_spec_index = None
-        for (rule_spec_index, rule_spec) in enumerate(answer_group.rule_specs):
-            evaluated_truth_value = rule_domain.evaluate_rule(
-                rule_spec, input_type, params, normalized_answer, fs)
-            if evaluated_truth_value > ored_truth_value:
-                ored_truth_value = evaluated_truth_value
-                best_rule_spec_index = rule_spec_index
-        if ored_truth_value > best_matched_truth_value:
+        ored_truth_value = rule_domain.CERTAIN_FALSE_VALUE
+        best_rule_spec = None
+        for (rule_spec_index, rule_spec) in enumerate(
+                answer_group.rule_specs):
+            if rule_spec.rule_type != rule_domain.FUZZY_RULE_TYPE:
+                evaluated_truth_value = rule_domain.evaluate_rule(
+                    rule_spec, input_type, params, normalized_answer, fs)
+                if evaluated_truth_value > ored_truth_value:
+                    ored_truth_value = evaluated_truth_value
+                    best_rule_spec_index = rule_spec_index
+                    best_rule_spec = rule_spec
+        if ored_truth_value == rule_domain.CERTAIN_TRUE_VALUE:
             best_matched_truth_value = ored_truth_value
             best_matched_answer_group = answer_group
             best_matched_answer_group_index = answer_group_index
             best_matched_rule_spec_index = best_rule_spec_index
+            matched_rule_type = 'hard'
```

Comment:
We may be able to discern between hard and soft/classifier classifications, but I don't think we can discern between soft and classifier classifications when looking at rule_spec_index. The reason for this is the same rule spec is used by both of those classifiers to determine whether it's a soft match or one that needs to run the classifier.

I have broken up reader.classify into separate functions that are run by the test individually to determine this instead.